### ISSUE TITLE
TAIL over certain sources may return a timestamp of either 0 or u64::MAX

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#6295]
### What version of Materialize are you using?

```
v0.7.2-dev (375d3a5a7)
```

### What was the issue?

There is a race condition when TAIL-ing ```VIEW AS VALUES``` sources -- the timestamp column in the TAIL is either 0 immediately after the view is created or u64::MAX afterwards. I think other types of sources also use u64::MAX which means they may also be susceptible to returning 0 initially.

### Is the issue reproducible? If so, please provide reproduction instructions.

The issue is best reproducible with -w 1 and goes away with large number of workers. It also goes away if a sleep is introduced between creating the source and tailing it.

To reproduce:

1. Take the testdrive from philip_stoev:testdrive_no_fetch_retry (https://github.com/MaterializeInc/materialize/pull/6294)

2. Use this test:

```
> CREATE MATERIALIZED VIEW v1 AS VALUES (1),(2),(3);

> BEGIN

> DECLARE c CURSOR FOR TAIL v1;

> FETCH 4 c WITH (timeout='10s');
18446744073709551615 1 1
18446744073709551615 1 2
18446744073709551615 1 3
```

3. Run in a loop until it fails:
 
```
while ../../target/release/testdrive fetch.td ; do :; done
...
fetch.td:7:1: error: non-matching rows: expected:
[["18446744073709551615", "1", "1"], ["18446744073709551615", "1", "2"], ["18446744073709551615", "1", "3"]]
got:
[["0", "1", "1"], ["0", "1", "2"], ["0", "1", "3"]]
Diff:
extra row: ["0", "1", "1"]
extra row: ["0", "1", "2"]
extra row: ["0", "1", "3"]
row missing: ["18446744073709551615", "1", "1"]
row missing: ["18446744073709551615", "1", "2"]
row missing: ["18446744073709551615", "1", "3"]

     |
   6 | 
   7 | > FETCH 4 c WITH (timeout='10s');
     | ^
```

In other words, the timestamp column of the TAIL is either 0 or u64::MAX

**Comment by mjibson** [comment]
This is known and working as intended for some confusing reasons. I think it occurs because sometimes the TAIL runs before or after the view is compacted, and TAIL will tell you about the first time it knows about, which changes based on compaction. I don't think users will hit this (because they won't TAIL a static view). We also might change the max timestamp selection in the future (but that won't change the "sometimes 0" behavior).

**Comment by mjibson** [comment]
Ok, I think my analysis above was flawed because TAIL doesn't use information from the `Frontiers` object. However, TAIL does use `determine_frontier` which differs from `determine_timestamp`. `determine_timestamp` (SELECT) will [error](https://github.com/MaterializeInc/materialize/blob/7470ce56a6beae78d47f295fc03bec287ac6b07d/src/coord/src/coord.rs#L2280) if it is asked about a source/index that doesn't have any timestamps, which is the 0 init thing describe above. TAIL uses `determine_frontier` which does not error in that case and will [happily](https://github.com/MaterializeInc/materialize/blob/7470ce56a6beae78d47f295fc03bec287ac6b07d/src/coord/src/coord.rs#L2341) attempt to start at time 0. We should perhaps change `determine_frontier` to error in the same case?

This doesn't answer another question, though: why are we returning data at time 0 at all? One of us needs to look into how static views work. Maybe the answer is just to make static views more deterministic, and the rest of the system is fine.
